### PR TITLE
Optimized std.datetime.systime.SysTime.fromISOString

### PR DESCRIPTION
**Pull request by JackStouffer**
So this rendition is a bit messier than normal because of the issue of using `find`. `indexOf` doesn't have a multiple `needle`s, so I tried to refactor to use `countUntil`, but it ended up being worse readability wise because you need to know which of the elements was found. 

The `static if`s are there because the return type of `byCodeUnit` is different if the source is an array of `dchar`s. In hindsight, this doesn't seem like this was the best idea but w/e.

No performance numbers for LDC because it doesn't have the change which gives access to `source` yet

```
$ dmd -O -inline -release test.d && ./test
2013-Feb-07 04:39:37.0000503
old		2 secs, 316 ms, 927 μs, and 3 hnsecs
new		1 sec, 350 ms, 355 μs, and 4 hnsecs
```

<details>

```d
import std.stdio;
import std.algorithm;
import std.exception;
import std.conv;
import std.ascii;
import std.range;
import std.traits;
import std.typecons;
import std.string;
import std.datetime.systime;
import std.datetime.date;
import std.datetime.timezone;
import std.datetime.stopwatch;
import core.time;
import std.utf;
import std.random;

enum testCount = 1_000_000;

__gshared a = "20130207T043937.000050392";

SysTime fromISOString1(S)(in S isoString, immutable TimeZone tz = null) @safe
        if (isSomeString!S)
{
    import std.algorithm.searching : startsWith, find;
    import std.conv : to;
    import std.string : strip;

    auto dstr = to!dstring(strip(isoString));
    immutable skipFirst = dstr.startsWith('+', '-') != 0;

    auto found = (skipFirst ? dstr[1..$] : dstr).find('.', 'Z', '+', '-');
    auto dateTimeStr = dstr[0 .. $ - found[0].length];

    dstring fracSecStr;
    dstring zoneStr;

    if (found[1] != 0)
    {
        if (found[1] == 1)
        {
            auto foundTZ = found[0].find('Z', '+', '-');

            if (foundTZ[1] != 0)
            {
                fracSecStr = found[0][0 .. $ - foundTZ[0].length];
                zoneStr = foundTZ[0];
            }
            else
                fracSecStr = found[0];
        }
        else
            zoneStr = found[0];
    }

    try
    {
        auto dateTime = DateTime.fromISOString(dateTimeStr);
        auto fracSec = fracSecsFromISOString(fracSecStr);
        Rebindable!(immutable TimeZone) parsedZone;

        if (zoneStr.empty)
            parsedZone = LocalTime();
        else if (zoneStr == "Z")
            parsedZone = UTC();
        else
        {
            try
                parsedZone = fromISOString(zoneStr);
            catch (DateTimeException dte)
                parsedZone = fromISOExtString(zoneStr);
        }

        auto retval = SysTime(dateTime, fracSec, parsedZone);

        if (tz !is null)
            retval.timezone = tz;

        return retval;
    }
    catch (DateTimeException dte)
        throw new DateTimeException(format("Invalid ISO String: %s", isoString));
}

SysTime fromISOString2(S)(in S isoString, immutable TimeZone tz = null) @safe
        if (isSomeString!S)
{
    import std.algorithm.searching : startsWith, find;
    import std.conv : to;
    import std.string : strip;
    import std.utf : byCodeUnit;

    auto str = strip(isoString);
    immutable skipFirst = str.startsWith('+', '-');

    auto found = (skipFirst ? str[1..$] : str).byCodeUnit.find('.', 'Z', '+', '-');
    auto dateTimeStr = str[0 .. $ - found[0].length];

    typeof(str) fracSecStr;
    typeof(str) zoneStr;

    if (found[1] != 0)
    {
        if (found[1] == 1)
        {
            auto foundTZ = found[0].find('Z', '+', '-');

            if (foundTZ[1] != 0)
            {
                static if (isNarrowString!S)
                {
                    fracSecStr = found[0][0 .. $ - foundTZ[0].length].source; 
                    zoneStr = foundTZ[0].source;
                }
                else
                {
                    fracSecStr = found[0][0 .. $ - foundTZ[0].length]; 
                    zoneStr = foundTZ[0];
                }
            }
            else
            {
                static if (isNarrowString!S)
                    fracSecStr = found[0].source;
                else
                    fracSecStr = found[0];
            }
        }
        else
        {
            static if (isNarrowString!S)
                zoneStr = found[0].source;
            else
                zoneStr = found[0];
        }
    }

    try
    {
        auto dateTime = DateTime.fromISOString(dateTimeStr);
        auto fracSec = fracSecsFromISOString(fracSecStr);

        Rebindable!(immutable TimeZone) parsedZone;

        if (zoneStr.empty)
            parsedZone = LocalTime();
        else if (zoneStr == "Z")
            parsedZone = UTC();
        else
        {
            try
                parsedZone = fromISOString(zoneStr);
            catch (DateTimeException dte)
                parsedZone = fromISOExtString(zoneStr);
        }

        auto retval = SysTime(dateTime, fracSec, parsedZone);

        if (tz !is null)
            retval.timezone = tz;

        return retval;
    }
    catch (DateTimeException dte)
        throw new DateTimeException(format("Invalid ISO String: %s", isoString));
}

Duration fracSecsFromISOString(S)(in S isoString) @trusted pure
if (isSomeString!S)
{
    import std.algorithm.searching : all;
    import std.ascii : isDigit;
    import std.conv : to;
    import std.string : representation;
    import core.time;

    if (isoString.empty)
        return Duration.zero;

    auto str = isoString.representation;

    enforce(str[0] == '.', new DateTimeException("Invalid ISO String"));
    str.popFront();

    enforce(!str.empty && all!isDigit(str), new DateTimeException("Invalid ISO String"));

    dchar[7] fullISOString = void;
    foreach (i, ref dchar c; fullISOString)
    {
        if (i < str.length)
            c = str[i];
        else
            c = '0';
    }

    return hnsecs(to!int(fullISOString[]));
}

immutable(SimpleTimeZone) fromISOString(S)(S isoString) @safe pure
        if (isSomeString!S)
{
    import std.algorithm.searching : startsWith;
    import std.conv : text, to, ConvException;
    import std.datetime.date : DateTimeException;
    import std.exception : enforce;

    auto whichSign = isoString.startsWith('-', '+');
    enforce!DateTimeException(whichSign > 0, text("Invalid ISO String ", isoString));

    isoString = isoString[1 .. $];
    auto sign = whichSign == 1 ? -1 : 1;
    int hours;
    int minutes;

    try
    {
        // cast to int from uint is used because it checks for
        // non digits without extra loops
        if (isoString.length == 2)
        {
            hours = cast(int) to!uint(isoString);
        }
        else if (isoString.length == 4)
        {
            hours = cast(int) to!uint(isoString[0 .. 2]);
            minutes = cast(int) to!uint(isoString[2 .. 4]);
        }
        else
        {
            throw new DateTimeException(text("Invalid ISO String ", isoString));
        }
    }
    catch (ConvException)
    {
        throw new DateTimeException(text("Invalid ISO String ", isoString));
    }

    enforce!DateTimeException(hours < 24 && minutes < 60, text("Invalid ISO String ", isoString));

    return new immutable SimpleTimeZone(sign * (dur!"hours"(hours) + dur!"minutes"(minutes)));
}

immutable(SimpleTimeZone) fromISOExtString(S)(S isoExtString) @safe pure
        if (isSomeString!S)
{
    import std.algorithm.searching : startsWith;
    import std.conv : ConvException, to;
    import std.datetime.date : DateTimeException;
    import std.exception : enforce;
    import std.format : format;
    import std.string : indexOf;

    auto whichSign = isoExtString.startsWith('-', '+');
    enforce!DateTimeException(whichSign > 0, format("Invalid ISO String: %s", isoExtString));
    auto sign = whichSign == 1 ? -1 : 1;

    isoExtString = isoExtString[1 .. $];
    enforce!DateTimeException(!isoExtString.empty, format("Invalid ISO String: %s", isoExtString));

    immutable colon = isoExtString.indexOf(':');
    S hoursStr;
    S minutesStr;
    int hours, minutes;

    if (colon != -1)
    {
        hoursStr = isoExtString[0 .. colon];
        minutesStr = isoExtString[colon + 1 .. $];
        enforce!DateTimeException(minutesStr.length == 2, format("Invalid ISO String: %s", isoExtString));
    }
    else
    {
        hoursStr = isoExtString;
    }

    enforce!DateTimeException(hoursStr.length == 2, format("Invalid ISO String: %s", isoExtString));

    try
    {
        // cast to int from uint is used because it checks for
        // non digits without extra loops
        hours = cast(int) to!uint(hoursStr);
        minutes = cast(int) (minutesStr.empty ? 0 : to!uint(minutesStr));
    }
    catch (ConvException)
    {
        throw new DateTimeException(format("Invalid ISO String: %s", isoExtString));
    }

    enforce!DateTimeException(hours < 24 && minutes < 60, format("Invalid ISO String: %s", isoExtString));

    return new immutable SimpleTimeZone(sign * (dur!"hours"(hours) + dur!"minutes"(minutes)));
}

void main()
{
    SysTime res;

    auto result = to!Duration(benchmark!(() => res = fromISOString1(a))(testCount)[0]);
    auto result2 = to!Duration(benchmark!(() => res = fromISOString2(a))(testCount)[0]);

    writeln(res); // force execution

    writeln("old", "\t\t", result);
    writeln("new", "\t\t", result2);
}
```

</details>

### CODE REVIEW

**Comment by JackStouffer** [comment]
@wilzbach The trouble is that `byCodeUnit` on a `dstring` returns `dstring`.

**Comment by wilzbach** [comment]
@JackStouffer this is what I'm talking about:

```diff
diff --git a/std/datetime/systime.d b/std/datetime/systime.d
index 52f777e95..3676db1f2 100644
--- a/std/datetime/systime.d
+++ b/std/datetime/systime.d
@@ -8105,8 +8105,8 @@ public:
         auto found = (skipFirst ? str[1..$] : str).byCodeUnit.find('.', 'Z', '+', '-');
         auto dateTimeStr = str[0 .. $ - found[0].length];
 
-        typeof(str) fracSecStr;
-        typeof(str) zoneStr;
+        typeof(found[0].source) fracSecStr;
+        typeof(found[0].source) zoneStr;
 
         if (found[1] != 0)
         {
@@ -8116,31 +8116,17 @@ public:
 
                 if (foundTZ[1] != 0)
                 {
-                    static if (isNarrowString!S)
-                    {
-                        fracSecStr = found[0][0 .. $ - foundTZ[0].length].source;
-                        zoneStr = foundTZ[0].source;
-                    }
-                    else
-                    {
-                        fracSecStr = found[0][0 .. $ - foundTZ[0].length];
-                        zoneStr = foundTZ[0];
-                    }
+                    fracSecStr = found[0].source[0 .. $ - foundTZ[0].length];
+                    zoneStr = foundTZ[0].source;
                 }
                 else
                 {
-                    static if (isNarrowString!S)
-                        fracSecStr = found[0].source;
-                    else
-                        fracSecStr = found[0];
+                    fracSecStr = found[0].source;
                 }
             }
             else
             {
-                static if (isNarrowString!S)
-                    zoneStr = found[0].source;
-                else
-                    zoneStr = found[0];
+                zoneStr = found[0].source;
             }
         }
 
diff --git a/std/utf.d b/std/utf.d
index 71681d368..58e67d43a 100644
--- a/std/utf.d
+++ b/std/utf.d
@@ -3345,12 +3345,22 @@ if (isAutodecodableString!R ||
     else static if (is(R : const dchar[]) && !__traits(hasMember, R, "empty") &&
                     !__traits(hasMember, R, "front") && !__traits(hasMember, R, "popFront"))
     {
-        return cast(StringTypeOf!R) r;
+        static struct ByCodeUnitImpl
+        {
+            StringTypeOf!R source;
+            alias source this;
+        }
+        return ByCodeUnitImpl(cast(StringTypeOf!R) r);
     }
     else
     {
         // byCodeUnit for ranges and dchar[] is a no-op
-        return r;
+        static struct ByCodeUnitImpl
+        {
+            R source;
+            alias source this;
+        }
+        return ByCodeUnitImpl(r);
     }
 }
```

but I currently don't have time to dig into why it's making the `std.utf` testsuite fail / whether we can actually do this. It does pass your PR tests though.

File: std/datetime/systime.d [hunk 1]
```diff
@@ -8097,15 +8097,16 @@ public:
         import std.algorithm.searching : startsWith, find;
         import std.conv : to;
         import std.string : strip;
+        import std.utf : byCodeUnit;
 
-        auto dstr = to!dstring(strip(isoString));
-        immutable skipFirst = dstr.startsWith('+', '-') != 0;
+        auto str = strip(isoString);
+        immutable skipFirst = str.startsWith('+', '-');
```
[the remark this answers is not in the document]
Comment (JackStouffer) [reply std/datetime/systime.d]: `skipOver` doesn't have a overload which accepts a template sequence.

**Comment by JackStouffer** [comment]
I don't think we should focus on the `byCodeUnit` problem at the moment because the problem of changing the return value in this way is something I don't think we should do so lightly.

**Comment by jmdavis** [comment]
It would break tons of code if `byCodeUnit` suddenly returned a wrapper struct for dstrings. Sure, your stuck with more `static if`s in order to strip of `byCodeUnitImpl` from narrow strings, but it works. Until a release or two back, not even that was possible, because it didn't have `source`. But in order to make it so that `source` worked for anything passed to `byCodeUnit`, you'd have to wrap all types, and that would break existing code. It's the same problem when you use `take` or `takeExactly` on different types of ranges where some support slicing and some don't.

**Comment by jmdavis** [comment]
My solution to the static ifs in one of my projects was to create a `stripBCU` function which returned the original range if it wasn't wrapped in a `ByCodeUnitImpl` and returned `source` if it was, but because `ByCodeUnitImpl` is a Voldemort type, you have to pass it the original type as a template argument so that it knows whether to strip off `ByCodeUnitImpl` (it can't look for `source`, since the original range might have a `source`). So, maybe we should break `ByCodeUnitImpl` out as `ByCodeUnit` so that it can be introspected (maybe even providing `stripBCU` in std.utf), but that's a separate issue. The static ifs work just fine. They're just a bit noisy, unfortunately.

File: std/datetime/systime.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply std/datetime/systime.d]: https://github.com/dlang/phobos/pull/6143

**Pull request merged**